accepted/0000-registry-scoped-keyfile-and-certfile.md

```diff
@@ -0,0 +1,84 @@
+   - 👎 values may need to be kept in sync with source of truth for key/cert
+   - 👎 need to set these dynamically if you need to push to multiple registries
+2. Specify `key` and `cert` in project `.npmrc`
+   - 👎 credentials may be inadvertently checked in
+   - 👎 file may be read by other local users (depending on permissions)
+   - 👎 file may need to be kept in sync with source of truth for key/cert
+   - 👎 need to change this file if you need to push to multiple registries
+3. Specify `key` and `cert` in `$HOME/.npmrc`
+   - 👎 file may be read by other local users (depending on permissions)
+   - 👎 file may need to be kept in sync with source of truth for key/cert
+   - 👎 need to change this file if you need to push to multiple registries
+4. Specify `key` and `cert` in a custom `.npmrc` specified via `npm_config_userconfig` / `--userconfig`
+   - 👎 file may be read by other users (depending on permissions)
+   - 👎 file may need to be kept in sync with source of truth for key/cert
+   - 👎 need to set `userconfig` dynamically if you need to push to multiple registries
+
+### Work around current `noAuth` / `ENEEDAUTH` checks:
+
+If your registry authenticates solely via `mTLS`, you may be able to work around it, but it gets complicated:
+
+- You can use something besides `npm publish` or invoke `libnpmpublish` yourself to get around the `ENEEDAUTH` check.
+- You can set some explicit/bogus credentials, but you might need to do some custom work to make your registry accept/ignore them.
+- If you are doing dual publishing, there are some gotchas:
+  - You can put scoped credentials in an `.npmrc` file, but that has similar issues as `cert`/`key` workarounds above.
+  - You can set top-level env vars like `npm_config_username` / `npm_config__password`, but that has similar issues as `npm_config_cert` / `npm_config_key` above.
+  - You can't set a scoped registry password/auth/authToken via env vars due to [this bug](https://github.com/npm/config/issues/64).
+
+## Implementation
+
+1. Resolve scoped `certfile`/`keyfile` options and use them, i.e.
+   1. **`npm-registry-fetch`**: Update [`getAuth` + `Auth`](https://github.com/npm/npm-registry-fetch/blob/main/lib/auth.js) to resolve any scoped `certfile` and `keyfile` options into corresponding `cert` and `key` properties, and update [`regFetch`](https://github.com/npm/npm-registry-fetch/blob/main/lib/index.js#L26) to use them in the `fetch` call.
+2. Accept `certfile`/`keyfile` as valid creds, i.e.
+   1. **`@npmcli/config`**: Update [`getCredentialsByURI`](https://github.com/npm/config/blob/1244177d8a68f68f2be46d5b9c21957da7dedfbb/lib/index.js#L759) to process scoped `certfile`/`keyfile` options and then set them in the returned credentials.
+   2. **`npm`**: Update [this `noCreds` check](https://github.com/npm/cli/blob/e992b4a21ecdd96aa33c59682c0ac0cc8a30d776/lib/commands/publish.js#L108) to accept `certfile`/`keyfile` as creds.
```

Comment:
Digging a little deeper, looks like [cli:lib/utils/get-identity.js](https://github.com/npm/cli/blob/ee3308a7a08799ec7e86237165ebaf278d9a4f9f/lib/utils/get-identity.js) and friends may need to be updated too, as there are some assumptions around it being possible to get the current username, which would not necessarily be the case if authn is strictly done via mTLS 🤔 . 

If that ends up being a whole can of worms, maybe this RFC could be reduced in scope (pun!) so as to cover just registry-scoped certfile/keyfile, leaving out the only-authenticate-via-mTLS feature. Then (I think) the changes can be limited to just npm-registry-fetch.

Reply:
IMO though it is reasonable to require custom registries to implement a `/-/whoami` for compatibility, in which case a meaningful username could be returned by `get-identity` in these cases so that other commands like `unpublish` can work solely via mTLS

Reply:
The "fail if you're not logged in" checks are a UX thing, and it's probably ok to err on the side of trying the attempt if we have a keyfile.